Added support for Marlin T0 redundant, probe and chamber temperatures

BToersche opened the pull request:
Changes in the Temperatures tab. The redundant and probe temperatures do not have controls associated with them, since they cannot be controlled after all. Each of the temperatures have been added to the 2 graphs at the bottom:
- Redundant: Grey line in extruder temp graph;
- Probe: Dark blue line in bed temp graph;
- Chamber: Black line in bed temp graph.
![image](https://user-images.githubusercontent.com/16536432/86850148-19a1cd00-c0b1-11ea-9af7-49edf10c4157.png)

Added options for enabling/disabling to the preferences menu (defaults are: "disabled").
![image](https://user-images.githubusercontent.com/16536432/86850157-1f97ae00-c0b1-11ea-934b-c366909478d1.png)

Some notes:
- Since I am not fluent in every language, I have only added the English translation for the newly added text. I did however add all new entries in each of the translation files and the template file.
- The redundant thermistor requires an "M105 R" (rather than a standard "M105") for receiving data, either an "M105 R" or "M105" is used depending on whether or not you have the redundant thermistor enabled in the interface.
- As shown in the top screenshot, the only connected thermistors are T0, redundant and bed. I have no probe and chamber thermistor installed, so I can't test these on my machine. I'll be receiving a probe within the next month though. I have however double checked Marlin's source (see `print_heater_state` in temperature.cpp at line 2868) to make sure the probe and chamber work properly.
- I have tested this on my machine which runs the latest version of Marlin (branch: bugfix-2.0.x) on an SKR1.4 board.

luc-github commented:
interresting I did not know the `M105 R` it is not documented in GCODE list : https://marlinfw.org/docs/gcode/M105.html

same for Probe and Chamber

are these controls compatible in repetier and smoothieware ?  I did not see any fw check

BToersche commented:
Rather than waiting for my probe to arrive, I've decided to test the code by editing the pins in my printer's firmware such that the probe and chamber temps would read a copy of the already installed probes. All works as it should:
![image](https://user-images.githubusercontent.com/16536432/86975300-54bb0380-c178-11ea-80d7-e85236894e55.png)

In terms of firmware compatibility, I've checked each of the firmwares to see which of the thermistors are supported:

Marlin
- Redundant	Yes	(see `print_heater_states` at line 2920 in temperature.cpp)
- Probe		Yes	(see `print_heater_states` at line 2950 in temperature.cpp)
- Chamber	Yes	(see `print_heater_states` at line 2937 in temperature.cpp)

Marlin Kimbra
- Redundant	No	(see `print_heaterstates` at line 3096 in MK_Main.cpp)
- Probe		No	(see `print_heaterstates` at line 3096 in MK_Main.cpp)
- Chamber	Yes	(see `print_chamberstate` at line 3140 in MK_Main.cpp)

Repetier (Both ArduinoAVR and ArduinoDUE)
- Redundant	No	(see `printTemperatures` at line 193 in Commands.cpp)
- Probe		No	(see `printTemperatures` at line 193 in Commands.cpp)
- Chamber	No*	(see `printTemperatures` at line 254 in Commands.cpp)

(*) Repertier is able to show chamber temperatures, however these are stub temperatures copied from extruder0.

Repetier for Davinci (Both ArduinoAVR and ArduinoDUE)
- Redundant	No	(see `printTemperatures` at line 155(AVR)/279(DUE) in Commands.cpp)
- Probe		No	(see `printTemperatures` at line 155(AVR)/279(DUE) in Commands.cpp)
- Chamber	No	(see `printTemperatures` at line 155(AVR)/279(DUE) in Commands.cpp)

Grbl
- Redundant	No	(see `gc_execute_line` at line 245 in gcode.c)
- Probe		No	(see `gc_execute_line` at line 245 in gcode.c)
- Chamber	No	(see `gc_execute_line` at line 245 in gcode.c)

Smoothieware
- Redundant	Yes*	(see `on_gcode_received` at line 229 in TemperatureControl.ccp)
- Probe		Yes*	(see `on_gcode_received` at line 229 in TemperatureControl.ccp)
- Chamber	Yes*	(see `on_gcode_received` at line 229 in TemperatureControl.ccp)

(*) By default, the configuration files only include hotend and bed `temperature_control`, however it seems you can simply add the required `temperature_control` modules, since you can change the designator (`R`/`P`/`C`, line 140), the GCode (to M141 for chamber, line 135) and not set a heater_pin for readonly mode (for redundant and probe, line 162). Though, this possibility might not immediately be clear for users.

For Smoothieware it might be a good idea to add some information next to the options, because people might not know the required designator for each of the thermistors. Of course this information is hidden when another firmware is used:
![image](https://user-images.githubusercontent.com/16536432/86974974-8f706c00-c177-11ea-95f1-a53dfb72998a.png)
Let me know if you agree.

I'll add the required firmware checks to the PR tomorrow or the day after.

luc-github commented:
No need to add smoothieware information in webUI should be ok to be in wiki

So please hide when not applicable in preferences - UI should follow
also there are marlin_embeded which act as marlin
and grbl_embeded which act as grbl 

I will see how to add these in 3.0
3.0 is more flexible than 2.X- it build UI according the content of M105 answer, and because in 3.0, the poilling  command is configurable by user, no need to have new preferences setting to choose M105 vs M105 R

luc-github commented:
Also looking at UI screenshot - redondant should not have target output if cannot be controled, or I missed something?
I guess probe should be same

BToersche commented:
I've added the required firmware checks for the temperature controls and tested these on my machine. Like stated, my machine runs Marlin, so I can only really test whether or not it works for the Marlin firmware. I did however also test to see if the options would be properly hidden if I removed Marlin from the supported firmwares, which worked correctly.
![image](https://user-images.githubusercontent.com/16536432/87225749-74784480-c38f-11ea-8ac4-feaa51ab1eec.png)

As for the target temperature values:
When a target is set for Heater T0, Marlin will also report this target temperature for the redundant sensor. This makes sense considering the redundant sensor is measuring the same heater as T0:
![image](https://user-images.githubusercontent.com/16536432/87225520-e8195200-c38d-11ea-9b13-d7c690ffa410.png)
As for the Probe, I intentionally kept the target temperature value in the UI as well because:
- It would be the only temperature sensor without a target value otherwise, which might look off in the GUI;
- And probe temperature readings are returned including a target temperature, might as well show it.


Also: I merged the base branch into this branch and rebuild the `index.html.gz` files including your changes.

luc-github commented:
Ok thank you for clarification - I will review on my side and merge it once done

luc-github commented:
I have added same support in 3.0 
![image](https://user-images.githubusercontent.com/8822552/87248174-427ee500-c458-11ea-8dc0-1746fccee2f1.png)

luc-github commented:
@all-contributors please add @BToersche code improvement
